controlplane/kubeadm/controllers/upgrade.go

```diff
@@ -64,38 +67,54 @@ func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(
 		return ctrl.Result{}, errors.Wrap(err, "failed to set role and role binding for kubeadm")
 	}
 
-	if err := workloadCluster.UpdateKubernetesVersionInKubeadmConfigMap(ctx, parsedVersion); err != nil {
+	configMapKey := ctrlclient.ObjectKey{Name: "kubeadm-config", Namespace: metav1.NamespaceSystem}
+	kubeadmConfig, err := workloadCluster.GetKubeadmConfig(ctx, configMapKey)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to fetch kubeadm config map")
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(kubeadmConfig.GetConfigMap(), r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
```

Comment:
@fabriziopandini @vincepri I was having a conversation with @shysank around this, and he mentioned that workload clusters **might be** managing their own config maps (and these config maps do not get stored in the management cluster). Is that actually the case, in which case, the client used to initialize the patchHelper above would not do the right thing.

Reply:
The kubeadm configmap should be technically only managed through a Cluster API deployment (KCP in this case), 

With the changes above, it doesn't seem to me that we'd overwrite other data, apart from the one that we're managing?

Reply:
So the patchHelper initialized with the `managementCluster.Client` would correctly persist the changes to the kubeadmConfig. Thanks for clearing it up.

Reply:
Just want to confirm on the patch helper client. The code above gets the config map from the workload cluster, but the patch helper updates the config map in the management cluster. Shouldn't the patch helper also use the workload cluster's client? I'm not entirely sure if we should be updating the management cluster's kubeadm config map.

Reply:
yes ^